ucb-rit/coldfront
# Refactor project user handling; make it easier to test BRC and LRC

**Changes**
- Moved logic to be run when a `User` is added to or joins a `Project` into abstract and concrete, deployment-specific runner classes.
- Refactored `ProjectAddUsersView` and `ProjectReviewJoinRequestsView` to instantiate a deployment-specific runner, perform changes in transactions, and reduce differences from the original repository.
- Introduced strategy classes for varying email behavior. Specifically, a class may choose to send an email immediately, enqueue it for later sending, or drop it entirely.
	- Used the second option to wait until the outermost transaction has committed before sending success emails.
- Rewrote `ProjectClusterAccessRequestRunner`.
	- Relaxed the requirement that the `Project`'s compute `Allocation` be active.
	- Wrapped `run` logic in a transaction.
- Moved code relating to join requests into dedicated directories.
- Added and updated tests.
	- Added a class `enable_deployment` that may be used as a context manager or decorator for enabling BRC and disabling LRC (or vice versa) in tests.
	    - See `coldfront.core.project.tests.test_utils.test_new_project_user_utils.test_runners` for example usage.

**How to Test**
- Review the code changes and add comments as needed.
- Ensure that the test suite passes.
	- Failures in secure directory tests in `coldfront.core.allocation.tests.test_views` are resolved in `develop`.
	- `coldfront.core.user.tests.test_views.test_request_hub_view.TestRequestHubView.test_cluster_access_requests` fails, but this is the case in the parent branch, `issue_337_cluster_access_api`.
- Manual Testing
	- Ensure that adding users, approving join requests, and processing cluster access requests behaves as expected.
